hmac/src/lib.rs

```diff
@@ -132,17 +97,35 @@ impl <D> Mac for Hmac<D>
         Self::new_varkey(key.as_slice()).unwrap()
     }
 
+    #[inline]
     fn new_varkey(key: &[u8]) -> Result<Self, InvalidKeyLength> {
+
         let mut hmac = Self {
             digest: Default::default(),
             i_key_pad: Default::default(),
             opad_digest: Default::default(),
-            key: expand_key::<D>(key),
+            key: Default::default(),
         };
-        hmac.i_key_pad = hmac.derive_key(IPAD);
+
+        if key.len() <= hmac.i_key_pad.len() {
+            hmac.i_key_pad[..key.len()].copy_from_slice(key);
+        } else {
+            let mut digest = D::default();
+            digest.process(key);
+            let output = digest.fixed_result();
+            hmac.i_key_pad[..output.len()].copy_from_slice(&output[..output.len()]);
+        }
+
+        let mut opad: GenericArray<u8, D::BlockSize> = GenericArray::default();
+
+        for (idx, elem) in hmac.i_key_pad.iter_mut().enumerate() {
+            *elem ^= IPAD;
+            opad[idx] = *elem ^ IOPAD;
+        }
```

Comment:
Is there a reason why not just use this variant?
```rust
for (ielem, oelem) in hmac.i_key_pad.iter_mut().zip(opad.iter_mut()) {
    *oelem = *ielem ^ OPAD;
    *ileem ^= IPAD;
}
```
I think this variant will be a bit clearer.

Reply:
It is much clearer. It would be interesting to know though if the `.zip` function that iterates simultaneously over to iterators is faster than a for loop?

Reply:
`zip` guarantees elimination of bounds checks, as it will iterate the minimum number of elements. It does not matter in this case, as `i_key_pad` and `opad` have the same size. Personally I am not sure if zip variant is better than the explicit counter:
```rust
let ipad = &mut hmac.i_key_pad;

for i in 0..Self::BlockSize::to_usize() {
    opad[i] = ipad[i] ^ OPAD;
    ipad[i] ^= IPAD;
}
```

But we better to stick one or another, and not mix them together.

Use of explicit counter is discouraged by clippy, but I think that in some cases it's more readable.